it/src/main/java/com/google/cloud/teleport/it/dataflow/DefaultDataflowClient.java

```diff
@@ -0,0 +1,399 @@
+          || matcher.find()) {
+        continue;
+      }
+
+      if (result.containsKey(metricName)) {
+        LOG.warn("Key {} already exists in metrics. Something might be wrong.", metricName);
+      }
+
+      if (metricUpdate.getScalar() != null) {
+        result.put(metricName, ((Number) metricUpdate.getScalar()).doubleValue());
+      } else if (metricUpdate.getDistribution() != null) {
+        // currently, reporting distribution metrics as 4 separate scalar metrics
+        ArrayMap distributionMap = (ArrayMap) metricUpdate.getDistribution();
+        result.put(metricName + "_COUNT", ((Number) distributionMap.get("count")).doubleValue());
+        result.put(metricName + "_MIN", ((Number) distributionMap.get("min")).doubleValue());
+        result.put(metricName + "_MAX", ((Number) distributionMap.get("max")).doubleValue());
+        result.put(metricName + "_SUM", ((Number) distributionMap.get("sum")).doubleValue());
+      } else if (metricUpdate.getGauge() != null) {
+        LOG.warn("Gauge metric {} cannot be handled.", metricName);
+        // not sure how to handle gauge metrics
+      }
+    }
+    return result;
+  }
+
+  /** Creates a JobInfo object from the provided parameters. */
+  private JobInfo getJobInfo(LaunchConfig options, JobState state, Job job, String runner) {
+    Map<String, String> labels = job.getLabels();
+    JobInfo.Builder builder =
+        JobInfo.builder()
+            .setJobId(job.getId())
+            .setCreateTime(job.getCreateTime())
+            .setSdk(job.getJobMetadata().getSdkVersion().getVersionDisplayName())
+            .setVersion(job.getJobMetadata().getSdkVersion().getVersion())
+            .setJobType(job.getType())
+            .setRunner(runner)
+            .setParameters(options.parameters())
+            .setState(state);
+    if (labels != null && !labels.isEmpty()) {
+      builder
+          .setTemplateType(
+              job.getLabels().getOrDefault("goog-dataflow-provided-template-type", null))
+          .setTemplateVersion(
+              job.getLabels().getOrDefault("goog-dataflow-provided-template-version", null))
+          .setTemplateName(
+              job.getLabels().getOrDefault("goog-dataflow-provided-template-name", null));
+    }
+    return builder.build();
+  }
+
+  /** Executes the specified command and parses the response to get the Job ID. */
+  private String executeCommandAndParseResponse(List<String> cmd) throws IOException {
+    Process process = new ProcessBuilder().command(cmd).redirectErrorStream(true).start();
+    String output = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
+    Matcher m = JOB_ID_PATTERN.matcher(output);
+    if (!m.find()) {
+      throw new RuntimeException(
+          String.format(
+              "Dataflow output in unexpected format. Failed to parse Dataflow Job ID. "
+                  + "Result from process: %s",
+              output));
+    }
+    String jobId = m.group(1);
```

Comment:
Prefer named groups in the regex 
```
  private static final Pattern JOB_ID_PATTERN = Pattern.compile("Submitted job: (?<jobId>\\S+)");
```

and then `m.group("jobId")`

Reply:
Done.